File: server/languages/cxx/source/CodegenManager.cpp
#include "CodegenManager.h"
#include "TempDirectory.h"
#include "Utility.h"

#include <boost/process.hpp>
namespace process = boost::process;

CodegenManager::CodegenManager(const string& templateDir)
{
	this->redirectHeader = Utility::readFile(templateDir + "/redirect.h");
	this->redirectCode = Utility::readFile(templateDir + "/redirect.cpp");
	this->templateCode = Utility::readFile(templateDir + "/template.cpp");
	this->cmakeLists = Utility::readFile(templateDir + "/CMakeLists.txt");
}

string CodegenManager::performCodegen(const string& prefixCode, const string& mainCode)
{
	//Create an auto-deleting temporary directory to hold our codegen build
	TempDirectory tempDir;
	
	//Populate the template source code
	string codegenSource = this->templateCode;
	codegenSource = Utility::strReplace("//$$__PREFIX_CODE__$$", prefixCode, codegenSource);
	codegenSource = Utility::strReplace("//$$__MAIN_CODE__$$", mainCode, codegenSource);
	
	//Write our template files to the temp directory
	Utility::writeFile(tempDir.getPath() + "/redirect.h", this->redirectHeader);
	Utility::writeFile(tempDir.getPath() + "/redirect.cpp", this->redirectCode);
	Utility::writeFile(tempDir.getPath() + "/template.cpp", codegenSource);
	Utility::writeFile(tempDir.getPath() + "/CMakeLists.txt", this->cmakeLists);
	
	//Attempt to perform codegen
	auto cmake = process::search_path("cmake");
	Utility::execute(tempDir.getPath(), cmake, ".");
	ProcessOutput output = Utility::execute(tempDir.getPath(), cmake, "--build", ".", "--target", "install");
	if (output.code != 0) {
		throw std::runtime_error("codegen failed with exit code " + std::to_string(output.code) + " and stdout: \"" + output.stdout + "\" and stderr: \"" + output.stderr + "\"");
	}
	
	//Retrieve the generated executable data
	return Utility::readFile(tempDir.getPath() + "/bin/out");
}

File: server/languages/csharp/source/CodegenManager.cs
using System;
using System.IO;
using System.IO.Compression;

namespace language_csharp
{
	public class CodegenError : System.Exception {
		public CodegenError(string message) : base(message) {}
	}
	
	public class CodegenManager
	{
		private string templateCode;
		private string projectFile;
		
		public CodegenManager(string templateDir)
		{
			this.templateCode = File.ReadAllText(Path.Combine(templateDir, "Program.cs.template"));
			this.projectFile = File.ReadAllText(Path.Combine(templateDir, "CodegenOutput.csproj.template"));
		}
		
		public byte[] performCodegen(string prefixCode, string mainCode)
		{
			using (TempDirectory tempDir = new TempDirectory())
			{
				//Populate the template source code
				string codegenSource = this.templateCode;
				codegenSource = codegenSource.Replace("//$$__PREFIX_CODE__$$", prefixCode);
				codegenSource = codegenSource.Replace("//$$__MAIN_CODE__$$", mainCode);
				
				//Write our template files to the temp directory
				File.WriteAllText(Path.Combine(tempDir.getPath(), "Program.cs"), codegenSource);
				File.WriteAllText(Path.Combine(tempDir.getPath(), "CodegenOutput.csproj"), this.projectFile);
				
				//Attempt to perform codegen
				ProcessOutput buildOutput = Utility.execute("dotnet", "publish -c Release -f netcoreapp2.0", tempDir.getPath());
				if (buildOutput.code != 0) {
					throw new CodegenError($"build failed with exit code: {buildOutput.code}, stdout: \"{buildOutput.stdout}\" and stderr: \"{buildOutput.stderr}\"");
				}
				
				//Attempt to zip up the codegen results
				string zipFile = Path.Combine(tempDir.getPath(), "zipped.zip");
				ZipFile.CreateFromDirectory(Path.Combine(tempDir.getPath(), "bin", "Release", "netcoreapp2.0", "publish"), zipFile);
				
				//Retrieve the generated zip file
				return File.ReadAllBytes(zipFile);
			}
		}
	}
}
